src/protocols/bdx/AsyncTransferFacilitator.h

```diff
@@ -0,0 +1,108 @@
+/*
+ *
+ *    Copyright (c) 2023 Project CHIP Authors
+ *
+ *    Licensed under the Apache License, Version 2.0 (the "License");
+ *    you may not use this file except in compliance with the License.
+ *    You may obtain a copy of the License at
+ *
+ *        http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *    Unless required by applicable law or agreed to in writing, software
+ *    distributed under the License is distributed on an "AS IS" BASIS,
+ *    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *    See the License for the specific language governing permissions and
+ *    limitations under the License.
+ */
+
+#include <lib/core/CHIPError.h>
+#include <lib/support/BitFlags.h>
+#include <messaging/ExchangeContext.h>
+#include <messaging/ExchangeHolder.h>
+#include <protocols/bdx/BdxTransferSession.h>
+
+#pragma once
+
+namespace chip {
+namespace bdx {
+
+/**
+ * An abstract class with methods for handling BDX messages from an ExchangeContext and using an event driven
+ * async approach to get the next action from the transfer session state machine.
+ *
+ * This class does not define any methods for beginning a transfer or initializing the underlying TransferSession object (see
+ * Responder below. TODO: # 29334 - Add Initiator to AsyncFacilitator)
+ *
+ * For each BDX transfer, we will have an instance of AsyncTransferFacilitator facilitating the transfer.
+ */
+class AsyncTransferFacilitator : public Messaging::ExchangeDelegate
+{
+public:
+    AsyncTransferFacilitator() : mExchange(*this) {}
+    ~AsyncTransferFacilitator() override = default;
+
+    /**
+     * This method should be implemented to contain business-logic handling of BDX messages and other TransferSession events.
+     *
+     * @param[in] event An OutputEvent that contains output from the TransferSession object.
+     */
+    virtual void HandleAsyncTransferSessionOutput(TransferSession::OutputEvent & event) = 0;
+
+    /**
+     * This method returns the exchange context contained in the exchange holder member of this class.
+     */
+    chip::Messaging::ExchangeContext * GetExchangeContext();
```

Comment:
So here is a question.  The _only_ consumer of this is `MTROTAImageTransferHandler::OnMessageToSend`.  But why are we asking our subclass to handle `kMsgToSend` events at all?  We have all the information we need to handle those....   What we do need to do is notify our subclass that the transfer is done if we hit errors there, right?  In general, we should think a bit about which event types this class should handle vs which ones the subclass should handle and how we do the transfer lifetime tracking.

Reply:
hmm. Thats a good idea. we could handle kmsgtosend from the base class itself and not pass the exchange context pointer around.